### ISSUE TITLE
type-intersection environment bug

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#23017]
The result is missing the `where T`
```julia
julia> z2(::Type{T} where T<:Tuple{Vararg{S}}) where {S} = S
z2 (generic function with 1 method)

julia> z2(Tuple{Vararg{Complex{T}}} where T)
Complex{T}
```

**Comment by vtjnash** [comment]
When this is fixed, the method
```
essentials.jl:69
convert(::Type{Tuple{Vararg{T}}}, x::Tuple) where {T} = cnvt_all(T, x...)
```
Should be re-written with the above form, so that it accepts any NTuple, and not just ones where N is unknown. For example, `Tuple{Complex, Vararg{Complex}}` and `Tuple{Complex, Complex}` should both also work with calling this method.

**Comment by JeffBezanson** [comment]
Yes, the free typevar there is definitely a bug.

We also have this behavior:

```
julia> Tuple{Complex,Complex} <: NTuple
false
```

Basically `NTuple` is a "diagonal" type. So I'm not sure those types would match this method.

**Comment by vtjnash** [comment]
Oops, that wasn't supposed to be a diagonal type, I copied the wrong example. The types should have been constrained to be the same: `Tuple{Complex{T}, Complex{T}} where T` like in the OP.

**Comment by JeffBezanson** [comment]
This is now a method error. Fixed by #23691 and #23643?
